<?xml version="1.0"?>
<launch>
  <!-- Gazebo & GUI Configuration -->
  <arg name="gazebo"      default="true"  doc="Should the gazebo simulation be launched? Use false in case if you want to include this file and launch gazebo yourself" />
  <arg name="headless"    default="true" doc="Should the gazebo GUI be launched?" />
  <arg name="paused"      default="false" doc="Should the simulation directly be stopped at 0s?" />
  <arg name="world"       default="worlds/empty.world" doc="Filename to a SDF World for gazebo to use" />
  <arg name="rviz"        default="false" doc="Should RVIz be launched?" />

  <!-- Robot Customization -->
  <arg name="arm_id"      default="panda" doc="Name of the robot to spawn" />
  <arg name="use_gripper" default="true"  doc="Should a franka hand be mounted on the flange?" />
  <arg name="controller"  default="shy_controller"     doc="Which example controller should be started? (One of {cartesian_impedance,model,force,joint_position,joint_velocity}_example_controller)" />
  <arg name="x"           default="0"     doc="How far forward to place the base of the robot in [m]?" />
  <arg name="y"           default="0"     doc="How far leftwards to place the base of the robot in [m]?" />
  <arg name="z"           default="0"     doc="How far upwards to place the base of the robot in [m]?" />
  <arg name="roll"        default="0"     doc="How much to rotate the base of the robot around its X-axis in [rad]?" />
  <arg name="pitch"       default="0"     doc="How much to rotate the base of the robot around its Y-axis in [rad]?" />
  <arg name="yaw"         default="0"     doc="How much to rotate the base of the robot around its Z-axis in [rad]?" />
  <arg name="xacro_args"  default=""      doc="Additional arguments to pass to panda.urdf.xacro" />
  <arg name="initial_joint_positions"
       doc="Initial joint configuration of the panda after FCI 'drive to start' program. Specify as a list of name/value pairs in form of '-J [name-of-joint] [value-in-rad]'. Default is a 90 degree bend in the elbow"
       default="-J $(arg arm_id)_joint1 0.104531
                -J $(arg arm_id)_joint2 -0.67887
                -J $(arg arm_id)_joint3 -0.08924
                -J $(arg arm_id)_joint4 -2.24591
                -J $(arg arm_id)_joint5 -0.01707
                -J $(arg arm_id)_joint6 1.645282
                -J $(arg arm_id)_joint7 0.896673
                -J $(arg arm_id)_finger_joint1 0.001
                -J $(arg arm_id)_finger_joint2 0.001"
       />
  <arg name="interactive_marker" default="False" doc="Should the interactive marker node be started?" />
  <!-- Topic interface in addition to actions -->
  <!-- <remap to="move_group/display_planned_path" from="/shy_controller/command" /> -->

  <include file="$(find franka_experiments)/launch/robot.launch">
    <arg name="gazebo" value="$(arg gazebo)" />
    <arg name="headless" value="$(arg headless)" />
    <arg name="paused" value="$(arg paused)" />
    <arg name="world" value="$(arg world)" />
    <arg name="rviz" value="$(arg rviz)" />

    <arg name="robot" value="panda" />
    <arg name="arm_id" value="$(arg arm_id)" />
    <arg name="use_gripper" value="$(arg use_gripper)" />
    <arg name="controller" value="$(arg controller)" />
    <arg name="x" value="$(arg x)" />
    <arg name="y" value="$(arg y)" />
    <arg name="z" value="$(arg z)" />
    <arg name="roll" value="$(arg roll)" />
    <arg name="pitch" value="$(arg pitch)" />
    <arg name="yaw" value="$(arg yaw)" />

    <arg name="xacro_args" value="$(arg xacro_args)" />
    <arg name="initial_joint_positions" value="$(arg initial_joint_positions)" />
    <arg name="interactive_marker" value="$(arg interactive_marker)" />

  </include>

  <rosparam file="$(find franka_experiments)/config/hw_controllers.yaml" command="load"  subst_value="true"/>
  <include file="$(dirname)/panda_moveit_controller_manager.launch.xml" />

  <rosparam command="load" file="$(find franka_experiments)/config/franka_example_controllers.yaml" subst_value="true" />

  <!-- as well as MoveIt demo -->
  <include file="$(find panda_moveit_config)/launch/demo.launch" pass_all_args="true">
    <!-- robot description is loaded by franka_control.launch -->
    <arg name="load_robot_description" value="false" />
    <!-- MoveItSimpleControllerManager provides ros_control's JointTrajectory controllers
         as well as GripperCommand actions -->
    <arg name="moveit_controller_manager" value="simple" />
  </include>

  <node name="rqt_reconfigure" pkg="rqt_reconfigure" type="rqt_reconfigure" required="false" />

</launch>
